### PR TITLE
Refactor large files into 1 file per class

### PR DESCRIPTION
**Pull request by ben-xD**
The main objective of this PR was to move classes (& enums) that were in the same file into their own separate file. Once I did that though, each file had to be exported via `ably_flutter.dart`, and that was not clean, `ably_flutter` would have had to contain a list of all files. The solution taken was to create separate mini libraries with their own `src` directory, and export the source files from these packages. Then inside `ably_flutter.dart`, these libraries will be exported instead of each file. I took the opportunity to move files into more logical folders. Other small changes are present: typo fix, unused import removal, simplify test teardown code.

This is a pattern observed in the [getx repo](https://github.com/jonataslaw/getx/tree/master/lib), though if another pattern is cleaner, please mention it.

This PR depends on https://github.com/ably/ably-flutter/pull/136 

**Notes for the future:** There are additional patterns we can learn/ follow in the future, and this refactor would help with this. Another improvement we can make is to create an internal platform library, and not expose these in the public API like we do currently. It also makes the transition to a full Dart SDK in the future easier too. For example, renaming `Rest` to `RestNative`, and changing the public API to rely on an [abstract factory](https://sourcemaking.com/design_patterns/abstract_factory). We can allow the user to use either the Native or the Dart SDK. 

I was worried about public API changes, and I don't think this PR affects it. As long as only public APIs/ classes are added to the library files (e.g. `ably_flutter.dart`, `push_notifications.dart`, `logging.dart`). Theres more about this in the [documentation](https://dart.dev/guides/libraries/create-library-packages): "Code under lib/src is considered private; other packages should never need to import src/.... To make APIs under lib/src public, you can export lib/src files from a file that’s directly under lib.", and on a [stackoverflow answer](https://stackoverflow.com/a/62487199/7365866).

### CODE REVIEW

**Comment by ben-xD** [comment]
Android 19 integration tests are failing, but after mentioning it in the stand up, we are actually dropping support for Android 19 and 20, raising minimum API level to 21. Please see [Review JDK 7 and Android API Level requirements for Java SDK](https://ably.atlassian.net/wiki/spaces/product/pages/1675362338/Review+JDK+7+and+Android+API+Level+requirements+for+Java+SDK) for more information.

**Review by QuintinWillison — APPROVED** [review]
Very difficult to review, given the size of change and amount of movement.

I think, in the interests of getting this landed sooner rather than later, it's best to rely on tests to indicate whether it still works. Also example app as proof of life.

**Review by tiholic — CHANGES_REQUESTED** [review]
Looks good overall, few comments:

File: lib/src/authentication/src/auth.dart [hunk 1]
```diff
@@ -1,5 +1,4 @@
-import 'common.dart';
-import 'rest/options.dart';
+import '../authentication.dart';
```
[the remark this answers is not in the document]
Comment (ben-xD) [reply lib/src/authentication/src/auth.dart:1]: I think the repetition of `src` is just the outcome of having modules inside modules whilst trying to avoid moving **every** file. Perhaps we can remove the first `src` folder from the file paths. WDYT?

This is what I meant by 
> This is a pattern observed in the [getx repo](https://github.com/jonataslaw/getx/tree/master/lib), though if another pattern is cleaner, please mention it.

File: lib/src/rest/src/options.dart [hunk 2]
```diff
@@ -0,0 +1 @@
+
```
[the remark this answers is not in the document]
Comment (ben-xD) [reply lib/src/rest/src/options.dart]: Removed in 0053a9ba0e8d854ba4e2872dc06f03214819e634

File: lib/src/stats/stats_internal_granularity.dart [hunk 3]
```diff
@@ -0,0 +1,14 @@
+/// https://docs.ably.com/client-lib-development-guide/features/#TS12c
```
[the remark this answers is not in the document]
Comment (ben-xD) [reply lib/src/stats/stats_internal_granularity.dart]: Renamed in 2a8c97b8c6c65bc005f6ff708d4e54cb024638b9, thanks

File: test/mock_method_call_manager.dart [hunk 4]
```diff
@@ -70,8 +67,9 @@ class MockMethodCallManager {
         // because function references (in `authCallback`) get dropped by the
         // PlatformChannel.
         if (!isAuthenticated && clientOptions.authUrl == 'hasAuthCallback') {
-          await AblyMethodCallHandler(methodChannel).onRealtimeAuthCallback(
-              AblyMessage(TokenParams(timestamp: DateTime.now()),
+          await AblyMethodCallHandler(Platform.methodChannel)
+              .onRealtimeAuthCallback(AblyMessage(
+                  TokenParams(timestamp: DateTime.now()),
                   handle: handle));
```
[the remark this answers is not in the document]
Comment (ben-xD) [reply test/mock_method_call_manager.dart]: Fixed in 9405f417392609df9a6c7168d00552f95c2fe832, thanks

File: lib/src/authentication/src/auth.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (ben-xD) [reply lib/src/authentication/src/auth.dart:1]: > Perhaps we can remove the first src folder from the file paths. WDYT?
I suggest we move the files after the PN PR is merged.

File: lib/src/platform/platform.dart [hunk 5]
```diff
@@ -0,0 +1,16 @@
+export 'src/ably_event_message.dart';
```
[the remark this answers is not in the document]
Comment (ben-xD) [reply lib/src/platform/platform.dart:1]: Originally, I would say there was still "implementation" inside the `spec` folder (the spec wasn't just abstract classes without implementation). I think the code in `impl` is specifically `platform` code, hence the rename. 

With the "mixing up", effectively, I moved all the `spec` files into the same level `impl` and renamed `impl` to `platform`. `platform` is still in its own folder/ module. So there is still a distinguishing feature, but less separate.

In federated plugins, the platform code would be in a separate package. 

Also, I think it would be easier to do file moves after the PN PR is merged (to avoid merge conflicts). We could consider moving all folder except `platform` into `non-platform`, because IMHO that is the only distinguishing feature between them. But maybe the folder names partly highlight this already?

File: lib/src/platform/src/platform.dart [hunk 6]
```diff
@@ -0,0 +1,61 @@
+import 'dart:async';
+
+import 'package:flutter/services.dart';
+
+import '../../error/error.dart';
+import '../../generated/platform_constants.dart';
+import '../platform.dart';
+
+class Platform {
```
[the remark this answers is not in the document]
Comment (ben-xD) [reply lib/src/platform/src/platform.dart:9]: I think it makes the code easier to read when using it/ importing, because importing the original file without a class will pollute the namespace with all the methods from the imported file. Now, they are all encapsulated in a `Platform` class. It is also easier to import, since Android Studio will not suggest importing a file `as Platform` (the class writer would know to `import ... as Platform`, but future users would just import directly using autoimport. 

TLDR: I was trying to remove the (arguably) _global_ methods, which is what the file originally had when writing code.

[the remark this answers is not in the document]
Comment (tiholic) [reply lib/src/platform/src/platform.dart:9]: Alright.

File: lib/src/platform/platform.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (tiholic) [reply lib/src/platform/platform.dart:1]: They do, but it is a little effortful to find out that `platform` is not a module similar to others (`push`, `rest`, `realtime`). 

I am unable to come up with any suggestion right now, but if needed, it's better to do after PN release.

File: lib/src/authentication/src/auth.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (tiholic) [reply lib/src/authentication/src/auth.dart:1]: > Perhaps we can remove the first src folder from the file paths

Sounds perfect.

> I suggest we move the files after the PN PR is merged.

Sure

**Review by tiholic — APPROVED** [review]
